### ISSUE TITLE
No way to create a server for Playwright tests

### ISSUE THREAD

**Issue by TheDutchCoder** [shunnNet/nuxt-msw#4]
We're not using vitest for our integration tests, but Playwright.
If we try to use `setupNuxtMswServer`, the module fails to load because it can't resolve `#imports` used in `@crazydos/nuxt-msw/test-utils`.

Is there a way to expose the MSW server for cases like these? Or will I just need to write my own custom MWS implementation to play with nuxt's `$fetch`?

The reason I need access to the server in the tests, is because we don't want to load all mocks globally all the time. We need fine-grained control over the implementations on a test level.

**Comment by shunnNet** [comment]
@TheDutchCoder 

Hello, thank you for raising this issue.

`setupNuxtMswServer` can only be used in the vitest environment nuxt(which has comment `// @vitest-environment nuxt`) as it depends on that specific environment, which is why it cannot be directly imported.

I’m not entirely sure how you’re running Playwright with Nuxt, so I’ll answer based on my understanding of Playwright.

Typically, when using `Playwright` for testing, it runs in an independent process, while `Nuxt` runs in another separate process. The `nuxt-msw` operates within the same process as `Nuxt`. Therefore, the process running `Playwright` cannot directly access the `MSW` server.

By the way, the description above also applies to [the way `@nuxt/test-utils` uses `Playwright`](https://nuxt.com/docs/getting-started/testing#end-to-end-testing).

To modify the `MSW` server in such scenarios, you can refer to MSW's suggestion for **dynamic mock scenarios**: https://mswjs.io/docs/best-practices/dynamic-mock-scenarios, which uses query parameters to change the handler state.

Since the configuration used by `nuxt-msw` in this case likely originates from `~/msw/server.ts`, you can access the `mswServer` and the `H3` event in the `onRequest` hook to retrieve current request information and modify the `MSW` server state. [Check doc here](https://github.com/shunnNet/nuxt-msw?tab=readme-ov-file#server)

One thing to note is that because the `MSW` server is always the same instance, if two tests run in parallel and both modify the server mocks, it may lead to unexpected bugs.

If the scenario I described above doesn’t match your situation, I might need more details, such as how you integrate `Playwright` and `Nuxt` for your tests.

Hope this helps!

**Comment by TheDutchCoder** [comment]
@shunnNet Thanks for the fast reply!

I'm currently using Playwright directly, since I can't get Nuxt to play nice with their own implementations of it (I think something's wrong in the setup method in the tests, but the error I get is nondescript, so hard to figure out).

Playwright and Nuxt run in separate processes indeed, the hard part is communicating between the Playwright test file and the Nuxt instance. That's why I attempted to setup a manual MSW implementation.

I have it all working with your module, but the issue I'm having is that I don't have access to the actual server in the Playwright test context.

I can try and see if I can get the vitest-way working, because I assume that would be ideal.

If you're on the Nuxt/Vueland Discord, then maybe we could bounce some ideas off each other?

**Comment by shunnNet** [comment]
@TheDutchCoder 
I'm on the Nuxt Discord under the name "shunnnet," so feel free to leave me a message. Due to time constraints, I might reply a bit later.

If it's possible to directly manipulate the MSW server from the Playwright script, that would indeed be ideal. Based on my search from over a year ago, there wasn't a straightforward solution for achieving this across processes back then (though perhaps there is one now). 

This is why I recommended using the approach provided by MSW's official documentation. While it's not direct, it's still acceptable. The method involves including query parameters when requesting the page URL.

One `nuxt-msw` related note is that in this scenario, you should avoid enabling `unit: true`; otherwise, it won't function properly.

**Comment by TheDutchCoder** [comment]
@shunnNet Thanks, I appreciate being available.

I actually _just_ got everything to work in vitest, so let me just tinker with it a bit more to see if I can get your module to now play nicely in that scenario.

I'll close the issue for now, as the original question is out of scope of the module.
